python/ray/serve/_private/proxy_request_response.py

```diff
@@ -160,28 +161,30 @@ def user_request(self) -> bytes:
     def send_request_id(self, request_id: str):
         self.context.set_trailing_metadata([("request_id", request_id)])
 
-    def send_status_code(self, status_code: grpc.StatusCode):
-        self.context.set_code(status_code)
-
-    def send_details(self, message: str):
-        self.context.set_details(message)
-
     def request_object(self, proxy_handle: ActorHandle) -> gRPCRequest:
         return gRPCRequest(
             grpc_user_request=self.user_request,
             grpc_proxy_handle=proxy_handle,
         )
 
 
-class ProxyResponse:
-    """ProxyResponse class to use in the common interface among proxies"""
+@dataclass(frozen=True)
+class ResponseStatus:
+    code: Any  # Must be convertible to a string.
+    is_error: bool = False
+    message: str = ""
 
-    def __init__(
-        self,
-        status_code: str,
-        response: Optional[bytes] = None,
-        streaming_response: Optional[Generator[bytes, None, None]] = None,
-    ):
-        self.status_code = status_code
-        self.response = response
-        self.streaming_response = streaming_response
+
+# Yields protocol-specific messages followed by a final `ResponseStatus`.
+ResponseGenerator = AsyncIterator[Union[Any, ResponseStatus]]
+
+
+@dataclass(frozen=True)
+class ResponseHandlerInfo:
+    response_generator: ResponseGenerator
+    application_name: str
+    deployment_name: str
+    route: str
```

Comment:
```suggestion
    application_name: str
    deployment_name: str
    route: str
```

Abstracting these into a standalone data-class (like `HandlerMetadata`) and making it optional would allow to: 
 - Avoid all the boilerplate above in the `_get_response_handler_info`
 - Avoid unused empty-field setting on the hot-path

Reply:
I feel those has different usecase and shouldn't be separated. Being explicit is also preferred over implicit in this case, else the logics will become hard to understand.

Reply:
How's HandlerMetadata going to make things implicit?

```

class HandlerResponseInfo:
  # ...
  metadata: Optional[HandlerMetadata]
```

Reply:
By making it optional and not setting them 😅

Reply:
We're not making it implicit -- these fields are not used in this path

Reply:
Hmm, I'm still not sure if I see the point, but will just make the change anyways if people this this is a better organization.